### ISSUE TITLE
aiochannel is stuck if we have move workers then items in channel

### ISSUE THREAD

**Issue by alvassin** [tbug/aiochannel#13]
Dear Henrik, i found the case, when aiochannel is getting stuck. Could you please have a look?

Long story short: if i remove context switch `await asyncio.sleep(0)` code works fine:

```python
# used python 3.10.2
import asyncio
from aiochannel import Channel

async def dummy_worker(ch):
    async for _ in ch:
        pass

async def main():
    concurrency = 2
    ch = Channel()
    tasks = [asyncio.create_task(dummy_worker(ch)) for _ in range(concurrency)]

    # Imagine we get items list asynchroniosly from web or moreover,
    # via async generator from stream and we do not know how items are there.
    # if i remove this context switch - everything works fine
    await asyncio.sleep(0)
    items = [1]

    for item in items:
        await ch.put(item)

    ch.close()
    await asyncio.gather(*tasks)


asyncio.run(main())
```

Two things to reproduce the bug:
1. Make context switch after worker tasks are created before first put
2. Do not make context switch after last put and before calling `channel.close()` method.

When calling `close()` method, [qsize is still not empty](https://github.com/tbug/aiochannel/blob/master/aiochannel/channel.py#L171), so one worker is not being cancelled:

```python
import asyncio
import logging

from aiochannel import Channel

async def dummy_worker(ch):
    async for _ in ch:
        pass

async def main():
    concurrency = 3
    ch = Channel()
    tasks = [asyncio.create_task(dummy_worker(ch)) for _ in range(concurrency)]

    # getters 0, qsize 0
    logging.info('getters %r, qsize %r', len(ch._getters), ch.qsize())

    await asyncio.sleep(0)

    # getters 3, qsize 0
    logging.info('getters %r, qsize %r', len(ch._getters), ch.qsize())
    items = [1]

    for item in items:
        await ch.put(item)

    # getters 2, qsize 1
    logging.info('getters %r, qsize %r', len(ch._getters), ch.qsize())

    ch.close()

    # getters 1, qsize 1
    logging.info('getters %r, qsize %r', len(ch._getters), ch.qsize())
    await asyncio.gather( *tasks)


logging.basicConfig(level=logging.INFO)
asyncio.run(main())
```

**Comment by tudborg** [comment]
I had a look @alvassin.

The underlying problem was that `close` did not wake/cancel up any remaining getters.

`close` will mark the channel as closed, but only cancel getters that _cannot_ be satisfied by the remaining items in the channel.

The problem is that nobody ever told the remaining getters that they could proceed, leading to hanging forever.

**Comment by tudborg** [comment]
A fix has been released to pypi in `v1.2.0`
